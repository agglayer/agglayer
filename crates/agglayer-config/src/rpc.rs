use std::{net::Ipv4Addr, str::FromStr, time::Duration};

use jsonrpsee::core::TEN_MB_SIZE_BYTES;
use serde::{Deserialize, Deserializer, Serialize};
use serde_with::serde_as;

/// The default port for the local RPC server.
const DEFAULT_PORT: u16 = 9090;

/// The local RPC server configuration.
#[serde_as]
#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct RpcConfig {
    /// If the `PORT` environment variable is set, it will take precedence over
    /// the configuration file.
    #[serde(default = "default_port", deserialize_with = "deserialize_port")]
    pub port: u16,
    #[serde(default = "default_host")]
    pub host: Ipv4Addr,

    /// The maximum size of the request body in bytes.
    #[serde(
        skip_serializing_if = "same_as_default_body_size",
        default = "default_body_size"
    )]
    pub max_request_body_size: u32,
    /// The maximum size of the response body in bytes.
    #[serde(
        skip_serializing_if = "same_as_default_body_size",
        default = "default_body_size"
    )]
    pub max_response_body_size: u32,
    /// The maximum number of connections.
    #[serde(
        skip_serializing_if = "same_as_default_max_connections",
        default = "default_max_connections"
    )]
    pub max_connections: u32,
    /// The maximum number of requests in a batch request. If `None`, the
    /// batch request limit is unlimited.
    #[serde(skip_serializing_if = "crate::default")]
    pub batch_request_limit: Option<u32>,
    /// The interval at which to send ping messages
    #[serde(skip)]
    #[serde_as(as = "Option<crate::with::HumanDuration>")]
    pub ping_interval: Option<Duration>,
    /// Timeout for completion of an RPC request to the AggLayer node.
    #[serde_as(as = "crate::with::HumanDuration")]
    #[serde(default = "default_request_timeout")]
    pub request_timeout: Duration,
}

impl Default for RpcConfig {
    fn default() -> Self {
        Self {
            port: default_port(),
            host: default_host(),
            max_request_body_size: default_body_size(),
            max_response_body_size: default_body_size(),
            max_connections: default_max_connections(),
            batch_request_limit: None,
            ping_interval: None,
            request_timeout: default_request_timeout(),
        }
    }
}

/// The default maximum number of connections.
fn default_max_connections() -> u32 {
    100
}

/// The default size of the request and response bodies in bytes.
fn default_body_size() -> u32 {
    TEN_MB_SIZE_BYTES
}

/// The default port for the local RPC server.
/// If the `PORT` environment variable is set, it will take precedence over
fn default_port() -> u16 {
    from_env_or_default("PORT", DEFAULT_PORT)
}

/// The default host for the local RPC server.
const fn default_host() -> Ipv4Addr {
    Ipv4Addr::new(0, 0, 0, 0)
}

/// Default timeout for completion of an RPC request to the AggLayer node.
const fn default_request_timeout() -> Duration {
    Duration::from_secs(180)
}

fn deserialize_port<'de, D>(deserializer: D) -> Result<u16, D::Error>
where
    D: Deserializer<'de>,
{
    let port = u16::deserialize(deserializer)?;

    Ok(from_env_or_default("PORT", port))
}

/// Get an environment variable or a default value if it is not set.
fn from_env_or_default<T: FromStr>(key: &str, default: T) -> T {
    std::env::var(key)
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or(default)
}

fn same_as_default_body_size(size: &u32) -> bool {
    *size == default_body_size()
}

fn same_as_default_max_connections(max_connections: &u32) -> bool {
    *max_connections == default_max_connections()
}
