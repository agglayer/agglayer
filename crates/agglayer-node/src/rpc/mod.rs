use std::sync::Arc;

use agglayer_config::epoch::BlockClockConfig;
use agglayer_config::Config;
use agglayer_config::Epoch;
use agglayer_storage::columns::latest_settled_certificate_per_network::SettledCertificate;
use agglayer_storage::stores::PendingCertificateReader;
use agglayer_storage::stores::PendingCertificateWriter;
use agglayer_storage::stores::StateReader;
use agglayer_storage::stores::StateWriter;
use agglayer_telemetry::KeyValue;
use agglayer_types::CertificateStatus;
use agglayer_types::EpochConfiguration;
use agglayer_types::{Certificate, CertificateHeader, CertificateId, Height, NetworkId};
use ethers::{
    contract::{ContractError, ContractRevert},
    providers::Middleware,
    types::H256,
};
use futures::TryFutureExt;
use jsonrpsee::{
    core::async_trait,
    proc_macros::rpc,
    server::{middleware::http::ProxyGetRequestLayer, PingConfig, ServerBuilder, ServerHandle},
};
use tokio::{sync::mpsc, try_join};
use tower_http::cors::CorsLayer;
use tracing::{debug, error, info, instrument};

use crate::{
    kernel::Kernel,
    rpc::error::{Error, RpcResult, StatusError},
    signed_tx::SignedTx,
};

mod error;
mod rpc_middleware;

#[cfg(test)]
mod tests;

#[rpc(server, namespace = "interop")]
trait Agglayer {
    #[method(name = "sendTx")]
    async fn send_tx(&self, tx: SignedTx) -> RpcResult<H256>;

    #[method(name = "getTxStatus")]
    async fn get_tx_status(&self, hash: H256) -> RpcResult<TxStatus>;

    #[method(name = "sendCertificate")]
    async fn send_certificate(&self, certificate: Certificate) -> RpcResult<CertificateId>;

    #[method(name = "getCertificateHeader")]
    async fn get_certificate_header(
        &self,
        certificate_id: CertificateId,
    ) -> RpcResult<CertificateHeader>;

    #[method(name = "getEpochConfiguration")]
    async fn get_epoch_configuration(&self) -> RpcResult<EpochConfiguration>;

    #[method(name = "getLatestKnownCertificateHeader")]
    async fn get_latest_known_certificate_header(
        &self,
        network_id: NetworkId,
    ) -> RpcResult<Option<CertificateHeader>>;
}

/// The RPC agglayer service implementation.
pub(crate) struct AgglayerImpl<Rpc, PendingStore, StateStore> {
    kernel: Kernel<Rpc>,
    certificate_sender: mpsc::Sender<(NetworkId, Height, CertificateId)>,
    pending_store: Arc<PendingStore>,
    state: Arc<StateStore>,
    config: Arc<Config>,
}

impl<Rpc, PendingStore, StateStore> AgglayerImpl<Rpc, PendingStore, StateStore> {
    /// Create an instance of the RPC agglayer service.
    pub(crate) fn new(
        kernel: Kernel<Rpc>,
        certificate_sender: mpsc::Sender<(NetworkId, Height, CertificateId)>,
        pending_store: Arc<PendingStore>,
        state: Arc<StateStore>,
        config: Arc<Config>,
    ) -> Self {
        Self {
            kernel,
            certificate_sender,
            pending_store,
            state,
            config,
        }
    }
}

impl<Rpc, PendingStore, StateStore> AgglayerImpl<Rpc, PendingStore, StateStore>
where
    Rpc: Middleware + 'static,
    PendingStore: PendingCertificateWriter + PendingCertificateReader + 'static,
    StateStore: StateReader + StateWriter + 'static,
{
    pub(crate) async fn start(self) -> anyhow::Result<ServerHandle> {
        // Create the RPC service
        let config = self.config.clone();
        let mut service = self.into_rpc();

        // Register the system_health method to serve health checks.
        service.register_method(
            "system_health",
            |_, _, _| serde_json::json!({ "health": true }),
        )?;

        // Create the RPC server.
        let mut server_builder = ServerBuilder::new()
            // Set the maximum request body size. The default is 10MB.
            .max_request_body_size(config.rpc.max_request_body_size)
            // Set the maximum response body size. The default is 10MB.
            .max_response_body_size(config.rpc.max_response_body_size)
            // Set the maximum number of connections. The default is 100.
            .max_connections(config.rpc.max_connections)
            // Set the batch request limit. The default is unlimited.
            .set_batch_request_config(match config.rpc.batch_request_limit {
                None => jsonrpsee::server::BatchRequestConfig::Unlimited,
                Some(0) => jsonrpsee::server::BatchRequestConfig::Disabled,
                Some(n) => jsonrpsee::server::BatchRequestConfig::Limit(n),
            });

        // Enable WebSocket ping/pong with the configured interval.
        // By default, pings are disabled.
        if let Some(duration) = config.rpc.ping_interval {
            server_builder =
                server_builder.enable_ws_ping(PingConfig::default().ping_interval(duration));
        }

        // Create a CORS middleware to allow cross-origin requests.
        let cors = CorsLayer::new()
            .allow_methods([
                hyper::Method::POST,
                hyper::Method::GET,
                hyper::Method::OPTIONS,
            ])
            .allow_origin(tower_http::cors::Any)
            .allow_headers([hyper::header::CONTENT_TYPE]);

        // Create a middleware stack with the CORS middleware and a proxy layer for
        // health checks.
        let middleware = tower::ServiceBuilder::new()
            .layer(ProxyGetRequestLayer::new("/health", "system_health")?)
            .layer(cors);

        let addr = config.rpc_addr();

        let server = server_builder
            .set_http_middleware(middleware)
            .set_rpc_middleware(rpc_middleware::from_config(&config))
            .build(addr)
            .await?;

        info!("Listening on {addr}");

        Ok(server.start(service))
    }
}

#[async_trait]
impl<Rpc, PendingStore, StateStore> AgglayerServer for AgglayerImpl<Rpc, PendingStore, StateStore>
where
    Rpc: Middleware + 'static,
    PendingStore: PendingCertificateWriter + PendingCertificateReader + 'static,
    StateStore: StateReader + StateWriter + 'static,
{
    #[instrument(skip(self, tx), fields(hash, rollup_id = tx.tx.rollup_id), level = "info")]
    async fn send_tx(&self, tx: SignedTx) -> RpcResult<H256> {
        let hash = format!("{:?}", tx.hash());
        tracing::Span::current().record("hash", &hash);

        info!(
            hash,
            "Received transaction {hash} for rollup {}", tx.tx.rollup_id
        );
        let rollup_id_str = tx.tx.rollup_id.to_string();
        let metrics_attrs = &[KeyValue::new("rollup_id", rollup_id_str)];

        agglayer_telemetry::SEND_TX.add(1, metrics_attrs);

        if !self.kernel.check_rollup_registered(tx.tx.rollup_id) {
            // Return an invalid params error if the rollup is not registered.
            return Err(Error::rollup_not_registered(tx.tx.rollup_id));
        }

        self.kernel.verify_signature(&tx).await.map_err(|e| {
            error!(error = %e, hash, "Failed to verify the signature of transaction {hash}: {e}");
            Error::signature_mismatch(e)
        })?;

        agglayer_telemetry::VERIFY_SIGNATURE.add(1, metrics_attrs);

        // Reserve a rate limiting slot.
        let guard = self
            .kernel
            .rate_limiter()
            .reserve_send_tx(tx.tx.rollup_id, tokio::time::Instant::now())?;

        agglayer_telemetry::CHECK_TX.add(1, metrics_attrs);

        // Run all the verification checks in parallel.
        try_join!(
            self.kernel
                .verify_proof_eth_call(&tx)
                .map_err(|e| {
                    let zkevm_error = decode_contract_error::<
                        _,
                        agglayer_contracts::polygon_zk_evm::PolygonZkEvmErrors,
                    >(&e);

                    let rollup_error = decode_contract_error::<
                        _,
                        agglayer_contracts::polygon_rollup_manager::PolygonRollupManagerErrors,
                    >(&e);

                    let error = match (zkevm_error, rollup_error) {
                        (Some(zkevm_error), _) => zkevm_error,
                        (_, Some(rollup_error)) => rollup_error,
                        (_, _) => e.to_string(),
                    };

                    error!(
                        error_code = %e,
                        error,
                        hash,
                        "Failed to dry-run the verify_batches_trusted_aggregator for transaction \
                         {hash}: {error}"
                    );
                    Error::dry_run(error)
                })
                .map_ok(|_| {
                    agglayer_telemetry::EXECUTE.add(1, metrics_attrs);
                }),
            self.kernel
                .verify_proof_zkevm_node(&tx)
                .map_err(|e| {
                    error!(
                        error = %e,
                        hash,
                        "Failed to verify the batch local_exit_root and state_root of transaction \
                         {hash}: {e}"
                    );
                    Error::root_verification(e)
                })
                .map_ok(|_| {
                    agglayer_telemetry::VERIFY_ZKP.add(1, metrics_attrs);
                })
        )?;

        // Settle the proof on-chain and return the transaction hash.
        let receipt = self.kernel.settle(&tx, guard).await.map_err(|e| {
            error!(
                error = %e,
                hash,
                "Failed to settle transaction {hash} on L1: {e}"
            );
            Error::settlement(e)
        })?;

        agglayer_telemetry::SETTLE.add(1, metrics_attrs);

        info!(hash, "Successfully settled transaction {hash}");

        Ok(receipt.transaction_hash)
    }

    #[instrument(skip(self), fields(hash = hash.to_string()), level = "info")]
    async fn get_tx_status(&self, hash: H256) -> RpcResult<TxStatus> {
        debug!("Received request to get transaction status for hash {hash}");
        let receipt = self.kernel.check_tx_status(hash).await.map_err(|e| {
            error!("Failed to get transaction status for hash {hash}: {e}");
            StatusError::tx_status(e)
        })?;

        let current_block = self.kernel.current_l1_block_height().await.map_err(|e| {
            error!("Failed to get current L1 block: {e}");
            StatusError::l1_block(e)
        })?;

        let receipt = receipt.ok_or_else(|| StatusError::tx_not_found(hash))?;

        let status = match receipt.block_number {
            Some(block_number) if block_number < current_block => "done",
            Some(_) => "pending",
            None => "not found",
        };
        Ok(status.to_string())
    }

    #[instrument(skip(self, certificate), fields(hash, rollup_id = *certificate.network_id), level = "info")]
    async fn send_certificate(&self, certificate: Certificate) -> RpcResult<CertificateId> {
        let hash = certificate.hash();
        tracing::Span::current().record("hash", hash.to_string());

        info!(
            %hash,
            "Received certificate {hash} for rollup {} at height {}", *certificate.network_id, certificate.height
        );

        // TODO: Batch the different queries.
        // Insert the certificate header into the state store.
        _ = self
            .state
            .insert_certificate_header(&certificate, CertificateStatus::Pending)
            .map_err(|e| {
                error!("Failed to insert certificate into state store: {e}");
                Error::internal(e.to_string())
            })?;

        // Insert the certificate into the pending store.
        _ = self
            .pending_store
            .insert_pending_certificate(certificate.network_id, certificate.height, &certificate)
            .map_err(|e| {
                error!("Failed to insert certificate into pending store: {e}");
                Error::internal(e.to_string())
            })?;

        if let Err(error) = self
            .certificate_sender
            .send((
                certificate.network_id,
                certificate.height,
                certificate.hash(),
            ))
            .await
        {
            error!("Failed to send certificate: {error}");

            return Err(Error::send_certificate(error));
        }

        Ok(hash)
    }

    async fn get_certificate_header(
        &self,
        certificate_id: CertificateId,
    ) -> RpcResult<CertificateHeader> {
        debug!("Received request to get certificate header for certificate {certificate_id}");
        match self.state.get_certificate_header(&certificate_id) {
            Ok(Some(header)) => Ok(header),
            Ok(None) => Err(Error::resource_not_found(format!(
                "Certificate({})",
                certificate_id
            ))),
            Err(error) => {
                error!("Failed to get certificate header: {}", error);

                Err(Error::internal("Unable to get certificate header"))
            }
        }
    }

    async fn get_epoch_configuration(&self) -> RpcResult<EpochConfiguration> {
        info!("Received request to get epoch configuration");

        if let Epoch::BlockClock(BlockClockConfig {
            epoch_duration,
            genesis_block,
        }) = self.config.epoch
        {
            Ok(EpochConfiguration {
                epoch_duration: epoch_duration.into(),
                genesis_block,
            })
        } else {
            Err(Error::internal(
                "AggLayer isn't configured with a BlockClock configuration, thus no \
                 EpochConfiguration is available",
            ))
        }
    }

    async fn get_latest_known_certificate_header(
        &self,
        network_id: NetworkId,
    ) -> RpcResult<Option<CertificateHeader>> {
        debug!(
            "Received request to get the latest known certificate header for rollup {}",
            network_id
        );

        let settled_certificate_id_and_height = self
            .state
            .get_latest_settled_certificate_per_network(&network_id)
            .map_err(|e| {
                error!("Failed to get latest settled certificate: {e}");

                Error::internal(e.to_string())
            })?
            .map(|(_, SettledCertificate(id, height, _, _))| (id, height));

        let proven_certificate_id_and_height = self
            .pending_store
            .get_latest_proven_certificate_per_network(&network_id)
            .map_err(|e| {
                error!("Failed to get latest proven certificate: {e}");
                Error::internal(e.to_string())
            })?
            .map(|(_, height, id)| (id, height));

        let pending_certificate_id_and_height = self
            .pending_store
            .get_latest_pending_certificate_per_network(&network_id)
            .map_err(|e| {
                error!("Failed to get latest pending certificate: {e}");
                Error::internal(e.to_string())
            })?
            .map(|c| (c.hash(), c.height));

        let certificate_id = [
            settled_certificate_id_and_height,
            proven_certificate_id_and_height,
            pending_certificate_id_and_height,
        ]
        .into_iter()
        .flatten()
        .max_by(|x, y| x.1.cmp(&y.1))
        .map(|v| v.0);

        if let Some(certificate_id) = certificate_id {
            match self.state.get_certificate_header(&certificate_id) {
                Ok(Some(header)) => Ok(Some(header)),
                Ok(None) => Err(Error::resource_not_found(format!(
                    "Certificate({})",
                    certificate_id
                ))),
                Err(error) => {
                    error!(
                        hash = certificate_id.to_string(),
                        "Failed to get certificate header: {}", error
                    );

                    Err(Error::internal("Unable to get certificate header"))
                }
            }
        } else {
            Ok(None)
        }
    }
}

fn decode_contract_error<M: Middleware, E: ContractRevert + std::fmt::Debug>(
    e: &ContractError<M>,
) -> Option<String> {
    e.decode_contract_revert::<E>()
        .map(|err| format!("{:?}", err))
}

type TxStatus = String;
