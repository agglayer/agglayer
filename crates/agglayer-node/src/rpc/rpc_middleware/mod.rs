//! RPC middleware layers.

use jsonrpsee::server::middleware::rpc::RpcServiceBuilder;
use tower::layer::util::{Identity, Stack};

mod logging_timeout;

pub use logging_timeout::LoggingTimeoutLayer;

/// The stack of RPC middleware layers.
pub type RpcStack = Stack<LoggingTimeoutLayer, Identity>;

/// Build the RPC middleware stack from config.
pub fn build(config: &agglayer_config::Config) -> RpcServiceBuilder<RpcStack> {
    jsonrpsee::server::middleware::rpc::RpcServiceBuilder::new()
        .layer(LoggingTimeoutLayer::new(config.rpc.request_timeout))
}
