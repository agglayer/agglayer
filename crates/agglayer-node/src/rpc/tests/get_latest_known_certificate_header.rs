use std::{sync::Arc, time::Duration};

use agglayer_config::Config;
use agglayer_storage::{
    storage::{pending_db_cf_definitions, state_db_cf_definitions, DB},
    stores::{pending::PendingStore, state::StateStore, PendingCertificateWriter, StateWriter},
    tests::TempDBDir,
};
use agglayer_types::{Certificate, CertificateHeader, CertificateStatus};
use insta::assert_snapshot;
use jsonrpsee::{core::client::ClientT, rpc_params};
use rstest::*;
use serde_json::json;

use super::TestContext;
use crate::rpc::AgglayerServer as _;

#[rstest]
#[test_log::test(tokio::test)]
async fn returns_the_pending_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);
    let proven_certificate = Certificate::new_for_test(network_id, 1);
    let pending_certificate = Certificate::new_for_test(network_id, 2);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&proven_certificate, CertificateStatus::Proven)
        .expect("unable to insert proven certificate header");
    state_db
        .insert_certificate_header(&pending_certificate, CertificateStatus::Pending)
        .expect("unable to insert pending certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &0,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    pending_db
        .set_latest_proven_certificate_per_network(&network_id, &1, &proven_certificate.hash())
        .expect("unable to set latest proven certificate");

    pending_db
        .insert_pending_certificate(network_id, 2, &pending_certificate)
        .expect("unable to insert pending certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, pending_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Pending);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::pending", json);
}

#[test_log::test(tokio::test)]
async fn returns_the_proven_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);
    let proven_certificate = Certificate::new_for_test(network_id, 1);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&proven_certificate, CertificateStatus::Proven)
        .expect("unable to insert proven certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &settled_certificate.height,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");
    pending_db
        .set_latest_proven_certificate_per_network(&network_id, &1, &proven_certificate.hash())
        .expect("unable to set latest proven certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, proven_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Proven);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::proven", json);
}

#[test_log::test(tokio::test)]
async fn returns_the_settled_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &settled_certificate.height,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, settled_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Settled);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!(
        "get_latest_known_certificate_header::settled_certificate",
        json
    );
}

#[test_log::test(tokio::test)]
async fn returns_no_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);

    let network_id = 1;

    let context = TestContext::new_with_config(config).await;

    let payload: Option<CertificateHeader> = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert!(payload.is_none());

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!(
        "get_latest_known_certificate_header::no_certificate_header",
        json
    );
}

#[test_log::test(tokio::test)]
async fn returns_the_highest_height() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 10);
    let pending_certificate = Certificate::new_for_test(network_id, 3);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&pending_certificate, CertificateStatus::Pending)
        .expect("unable to insert pending certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &10,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    pending_db
        .insert_pending_certificate(network_id, 3, &pending_certificate)
        .expect("unable to insert pending certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, settled_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Settled);
    assert_eq!(payload.height, 10);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::highest_height", json);
}
