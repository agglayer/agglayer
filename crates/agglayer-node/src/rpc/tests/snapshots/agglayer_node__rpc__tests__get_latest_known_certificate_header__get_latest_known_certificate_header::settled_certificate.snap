---
source: crates/agglayer-node/src/rpc/tests/get_latest_known_certificate_header.rs
expression: json
---
{
  "id": 0,
  "jsonrpc": "2.0",
  "result": {
    "certificate_id": "0xa68dc90fbf6055e0c747b63410172bb615b973ae65ec3bba93197f0282ef37e6",
    "certificate_index": null,
    "epoch_number": null,
    "height": 0,
    "metadata": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "network_id": 1,
    "new_local_exit_root": "0x0101010101010101010101010101010101010101010101010101010101010101",
    "prev_local_exit_root": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "status": "Settled"
  }
}
